Redesign right click menu system.

chairbender opened the pull request:
### Purpose
Reworks how the right click options work, allowing them to change dynamically and trying to improve code organization.
Fixes #1725 

### Open Questions and Pre-Merge TODOs

- [x]  This fix is tested on the same branch it is PR'ed to.
- [x]  I correctly commented my code
- [x]  My code is indented with tabs and not spaces
- [x]  This PR does not include any unnecessary .meta, .prefab or .unity (scene) changes
- [x]  This PR does not bring up any new compile errors
- [x]  This PR has been tested in editor
- [x]  This PR has been tested in multiplayer (with 2 clients and late joiners, if applicable)

### Notes:
Refer to the documentation on how this works here https://github.com/unitystation/unitystation/wiki/Right-Click-Menu-(Interaction-Framework-2)

As an example of how it can work dynamically, I made it so that all currently defined menu options should only be visible if you can actually perform the action. For example, when pulling an object, the Pull option goes away and is replaced with a Stop Pulling option.

chairbender committed:
#1725 Redesign right click menu system.

See https://github.com/unitystation/unitystation/wiki/Right-Click-Menu-(Interaction-Framework-2)

chairbender commented on UnityProject/Assets/Scripts/Electricity/NodeModules/BatterySupplyingModule.cs:
```diff
@@ -120,24 +120,12 @@ public override void PowerNetworkUpdate()
 		return (Current);
 	}
-
-	[ContextMethod("Toggle Charge", "Power_Button")]
```
oops, this shouldn't be deleted, will fix it

chairbender replied:
fixed

chairbender committed:
#1725 add missing options

chairbender committed:
#1725 refactor right click system to improve dev usability, WIP

chairbender committed:
#1725 finish improvements to right click system developer usability, also
allowing for customization!

Bod9001 commented on UnityProject/Assets/Scripts/UI/Right click/RightClickableResult.cs:
```diff
@@ -0,0 +1,91 @@
+	/// <param name="spriteOverride">Sprite to use instead of the RightClickOption's normal sprite.</param>
+	/// <param name="nameOverride">Name to use instead of the RightClickOption's name</param>
+	/// <returns>this RightClickableResult, to support chaining</returns>
+	public RightClickableResult AddElement(string optionName, Action action,
```
needs to have options for background sprite

Bod9001 replied:
There is some debate if we really need the custom backgrounds

My arguments for are, not too much to implement, it makes the UI look nice Once you've implemented it and it was in the previous system

Example of how it would be used
like with the crummy hive mind mode There is, you could right click and the background could be a swirly thing for any hive mind related action

yeah, it's up to you if you want to add it or not, is your choice

Bod9001 commented on UnityProject/Assets/Resources/ScriptableObjects/Interaction/RightclickOptions/PickUp.asset:
```diff
@@ -0,0 +1,17 @@
+  m_EditorClassIdentifier: 
+  icon: {fileID: 21300000, guid: 39da8ec811737ba4a92391a15a6c0f5b, type: 3}
+  label: Pick Up
+  backgroundColor: {r: 1, g: 1, b: 1, a: 1}
```
this is the wrong default background colour for sub buttons, it should be Gray, if I remember

Bod9001 reviewed (changes requested):
What I said in the comment, sadly am too dumb to put it in the right place xD

chairbender committed:
#1725 fix additional right click issues, add ability to set
background sprite

Bod9001 commented on UnityProject/Assets/Scripts/UI/Right click/RadialMenu.cs:
```diff
@@ -77,16 +77,12 @@ public class RadialMenu : MonoBehaviour {
-			newButton.Item = Menus [i].Item;
+			newButton.Circle.color = Menus[i].BackgroundColor;
+			newButton.Icon.sprite = Menus[i].IconSprite;
+			newButton.Background.sprite = Menus[i].BackgroundSprite;
```
it's Circle not Background, confusing I know right : P

Bod9001 commented on UnityProject/Assets/Scripts/UI/Right click/Button.prefab:
```diff
@@ -174,12 +284,58 @@ MonoBehaviour:
+  - component: {fileID: 2655520020073026784}
+  - component: {fileID: 7175192079277077561}
+  m_Layer: 5
+  m_Name: BGPanel
```
You didn't need to add this panel , since the circle is technically the background, You modify the sprite of the circle

Bod9001 reviewed (changes requested):
yeah, sorry from my vague naming of stuff, 

You meant to modify the circles sprite 
"newButton.Circle.sprite = Menus[i].BackgroundSprite;"
and there is no default background sprite set for the SOs 
You don't need the extra background panel, since you're just modifying the circle sprite 

It would be nice to change the background dynamically but it's up to you if you want to add that

You forgot to remove bob on the the cable and details is also missing

Nearly there ^^

chairbender committed:
#1725 fix errors related to inheritance when using
RightClickMethod, remove Bob method, fix background sprite
to replace circle when specified
